controllers/helmchart_controller.go

```diff
@@ -426,16 +435,147 @@ func (r *HelmChartReconciler) reconcileFromTarballArtifact(ctx context.Context,
 		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 	}
 
-	// Read the chart metadata
-	chartMetadata, err := chartutil.LoadChartfile(path.Join(chartPath, chartutil.ChartfileName))
+	// Read the chart
+	helmChart, err := loader.LoadDir(chartPath)
+	if err != nil {
+		err = fmt.Errorf("load chart error: %w", err)
+		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+	}
+
+	// Load dependencies
+	deps := helmChart.Metadata.Dependencies
+	lock := helmChart.Lock
+	if lock != nil {
+		// Load from lockfile if exists
+		deps = lock.Dependencies
+	}
+
+	// Create charts directory inside chart path
+	depChartsPath := path.Join(chartPath, "charts")
+	err = os.MkdirAll(depChartsPath, 0755)
 	if err != nil {
-		err = fmt.Errorf("load chart metadata error: %w", err)
+		err = fmt.Errorf("unable to create dependencies charts directory: %w", err)
 		return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 	}
 
+	for _, dep := range deps {
```

Comment:
Would probably be nice to use goroutines here and parallelize dependency sourcing.

Reply:
Yes, indeed.

In addition, I think it would be great if we could move most of this logic to `internal/helm` (but nothing directly related to working with our own Kubernetes resources, so we have a nice separation of concerns). This will likely mean that we first attempt to collect all our `HelmRepository` resources in the controller, and then construct _something_ (like the `downloader.Manager`) that is able to pull off the dependency collection job based on this generic information.

This also means that the core logic will be easier to test (as I am personally not to happy with the fat end-to-end tests we currently have).

Reply:
I've addressed all of the issues except for this which requires a bit more substantial intervention and refactoring. Will work on this next.